test/cpp_api_parity/functional_impl_check.py

```diff
@@ -0,0 +1,239 @@
+# The purpose of this test is to check that we have implementation parity between
+# a Python `torch.nn.functional` function and its corresponding C++ `torch::nn::functional`
+# function. Concretely, this test does the following:
+#
+# 1. Get a test params dict from common_nn.py, run forward pass on the Python functional
+# created using the test params.
+#
+# 2. Serialize the Python functional's forward input arguments, deserialize them
+# in C++ and use them as input for the C++ functional's forward pass.
+#
+# 3. Run the forward pass on the C++ functional, and serialize the C++ functional's
+# forward output.
+#
+# 4. Compare Python/C++ functional's forward output. If they are the same, then we
+# have implementation parity between Python/C++ module.
+
+import tempfile
+import shutil
+from string import Template
+import re
+import os
+
+import torch
+from cpp_api_parity.utils import TorchNNFunctionalTestParams, TORCH_NN_COMMON_TEST_HARNESS, \
+    compile_cpp_code_inline, set_python_tensors_requires_grad, move_python_tensors_to_device, \
+    add_test, compute_cpp_args_construction_stmts_and_forward_arg_symbols, serialize_arg_dict_as_script_module, \
+    compute_arg_dict, decorate_test_fn, compute_temp_file_path, generate_error_msg
+import cpp_api_parity
+
+# Expected substitutions:
+#
+# ${functional_variant_name}  (e.g. `BCELoss_no_reduce`)
+# ${cpp_args_construction_stmts}
+# ${cpp_function_call}
+TORCH_NN_FUNCTIONAL_TEST_FORWARD = Template("""
+void ${functional_variant_name}_test_forward(
+    const std::string& arg_dict_file_path,
+    const std::string& forward_output_file_path) {
+  pybind11::gil_scoped_release no_gil;
+
+  namespace F = torch::nn::functional;
+
+  // Declare arguments
+  auto arg_dict = load_dict_from_file(arg_dict_file_path);
+  ${cpp_args_construction_stmts};
+
+  // Some functionals (such as `F::rrelu`) create random tensors in their call path.
+  // To make sure the random tensors created are the same in Python/C++, we need
+  // to set the RNG seed manually.
+  torch::manual_seed(0);
+
+  // Run function with arguments
+  auto cpp_output = ${cpp_function_call};
+
+  // Save the output into a file to be compared in Python later
+  write_ivalue_to_file(torch::IValue(cpp_output), forward_output_file_path);
+}
+""")
+
+def run_forward(unit_test_class, test_params):
+    device = test_params.device
+
+    inputs = set_python_tensors_requires_grad([arg_value for _, arg_value in test_params.arg_dict['input']])
+    inputs = inputs + [arg_value for _, arg_value in test_params.arg_dict['target']]
+    inputs = inputs + [arg_value for _, arg_value in test_params.arg_dict['extra_args']]
+    inputs = move_python_tensors_to_device(inputs, device)
```

Comment:
Is it intentional that we're setting the requires grad to be True before we move the tensors onto the device?

Reply:
Thanks a lot for the catch! No it's not intentional, and I just changed their order to always move to device first before setting requires grad